### PR TITLE
Add reusable stress test class

### PR DESCRIPTION
**Pull request by reyang**
This was originally brought up while writing the lock-free circular buffer #1086.
The code was later copied and used in other places (e.g. #2330).

In this PR I'm trying to explore a way that we can encourage the maximum code reuse.

Notes for code reviewers: Please note that most of the decisions (e.g. why partial class, why static, etc.) were coming from perf - currently it only consumes 3 CPU cycles per loop on my machine. If you have other ideas, please get the perf numbers and see the diff.

```
E:\opentelemetry-dotnet\test\OpenTelemetry.Tests.Stress>dotnet run --framework netcoreapp3.1 --configuration Release
Running (concurrency = 1), press <Esc> to stop...
Loops: 1,525,149,797, Loops/Second: 699,946,926, CPU Cycles/Loop: 3
```

### CODE REVIEW

**Review by CodeBlanch — APPROVED** [review]
LGTM

I think it is very unusual to see this done through symlink (effectively) of Skeleton.cs and a `partial` Program class, but I don't have an issue with that. We could probably do it more like BenchmarkDotNet with a runner class + a bit of reflection to find the stress method to run but if this works, great, it isn't like we're shipping this.

File: test/StressTestMetrics/Program.cs [hunk 1]
```diff
@@ -15,84 +15,44 @@
...
-    {
-        private static readonly Meter MyMeter = new Meter("TestMeter", "0.0.1");
-        private static readonly Counter<long> Counter = MyMeter.CreateCounter<long>("counter");
-        private static string[] dimensionValues = new string[] { "DimVal1", "DimVal2", "DimVal3", "DimVal4", "DimVal5", "DimVal6", "DimVal7", "DimVal8", "DimVal9", "DimVal10" };
-        private static Random random = new Random();
+    private const int ArraySize = 10;
+    private static readonly Meter TestMeter = new Meter("TestMeter", "1.0.0");
+    private static readonly Counter<long> TestCounter = TestMeter.CreateCounter<long>("TestCounter");
+    private static readonly string[] DimensionValues = new string[ArraySize];
+    private static readonly ThreadLocal<Random> ThreadLocalRandom = new ThreadLocal<Random>(() => new Random());
```
Comment (cijothomas) [inline test/StressTestMetrics/Program.cs:33]: thanks for fixing threadsafety on Random!

**Review by cijothomas — APPROVED** [review]

File: test/OpenTelemetry.Tests.Stress/Skeleton.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                    Console.Title = output;
+                }
+            }, () =>
+            {
+                Parallel.For(0, concurrency, (i) =>
+                {
+                    statistics[i] = 0;
+                    while (bContinue)
+                    {
+                        statistics[i]++;
```
Comment (ThomsonTan) [inline test/OpenTelemetry.Tests.Stress/Skeleton.cs:104]: Do we need change the increment and above read to atomic?

**Comment by ThomsonTan** [comment]
BTW, I get 3 cycles per loop report on my laptop based on the default run (concurrent == 1 and release build), but there seems 7 instructions in the loop. So it looks like the processor runs ~2 instructions per cycle on a single core.

```
00007ffe`adfaa8d0 488bd0          mov     rdx,rax
00007ffe`adfaa8d3 3b7208          cmp     esi,dword ptr [rdx+8]
00007ffe`adfaa8d6 7319            jae     OpenTelemetry_Tests_Stress!Program+<>c__DisplayClass4_0.<Stress>b__2(Int32)+0x61 (00007ffe`adfaa8f1)
00007ffe`adfaa8d8 488d54da10      lea     rdx,[rdx+rbx*8+10h]
00007ffe`adfaa8dd 48ff02          inc     qword ptr [rdx]
00007ffe`adfaa8e0 803d3453090000  cmp     byte ptr [00007ffe`ae03fc1b],0
00007ffe`adfaa8e7 75e7            jne     OpenTelemetry_Tests_Stress!Program+<>c__DisplayClass4_0.<Stress>b__2(Int32)+0x40 (00007ffe`adfaa8d0)
```

File: test/OpenTelemetry.Tests.Stress/Skeleton.cs (hunk 2 above)
Comment (reyang) [reply test/OpenTelemetry.Tests.Stress/Skeleton.cs:104]: Why?

**Pull request merged**